### DISCUSSION TITLE
Custom build FreeType2 and HarfBuzz

### DISCUSSION THREAD

**Discussion by data-man** [contour-terminal/contour#291] [Development]
And link them statically. I think it will improve overall performance.

**Comment by christianparpart** [comment]
> And link them statically. I think it will improve overall performance.

Just for the performance reason? Or do you have any other argument in mind?

wrt performance, we have two main hot paths.

1. the rendering and input event loop (main thread)
2. the terminal processing unit (background thread)

The actual workload is done in the TPU thread, where *definitely* is still room for improvement.
On the main thread's rendering side I think it is (for now) more important to see if we can improve on OpenGL render times, such as making use of EBOs (research on whether or not that does indeed improve performance).

When some numbers clearly show a difference between statically linked FT/HB vs a dynamically linked one, I am all in - as long as we can make it cmake compile-time configurable what is to be used, because distributors prefer to use system dependencies instead of static links, as this makes updating for CVEs and bugfixes much easier.

**Reply by christianparpart** [reply]
@data-man deep down in my innocent head, I think we can at least match bandwidth throughput performance of Alacritty. Mind, in my personal bandwidth throughput benchmarks the ranking currently is: `Alacritty > Kitty > Contour > Gnome-Terminal > ....` (simple `cat`-style hugefile outputs). There are certainly more sophisticated tests that I definitely want to explore and would like to have integrated into the CI for pure statistical reasons. Let's see.

**Reply by data-man** [reply]
> Just for the performance reason? Or do you have any other argument in mind?

Yes and no.  I like rolling software (I'm Artix Linux user). :smile:
I just know that HarfBuzz can use decompressed Unicode data and I think it will improve performance.

**Reply by christianparpart** [reply]
I think it's totally fine to make it rolling via `cmake -D....`, as long as CI and release builds are deterministic.

**Reply by christianparpart** [reply]
On the HarfBuzz/FT matter, I am open to it, but let's wait with a bit until it does make sense to optimize into that direction? (as mentioned earlier, I think there're lower hanging fruids that could be found via `callgrind`.)

p.s.: first I need to shoot out the 0.2.0 release :D
